### ISSUE TITLE
quaternion_exp_to_log and euler_from_quaternion silently accept non-unit quaternions and return wrong values, unlike their scale-safe siblings

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#3953]
Two quaternion consumers in `kornia/geometry/conversions.py` assume their input is a unit quaternion but never normalize it and never check, so a non-unit quaternion silently produces a wrong answer. Their siblings in the same file are scale-safe, so the asymmetry is invisible from the call site and undocumented in every docstring involved.

- **`quaternion_exp_to_log`** — `quaternion_exp_to_log([0.5, 0.5, 0, 0])` returns `[1.0471975511965976, 0, 0]`, 33 % larger than the correct `[0.7853981633974484, 0, 0]`. Worse, `w > 1` is silently clamped: `quaternion_exp_to_log([2, 0, 0, 0])` returns the origin because `torch.clamp(quaternion_scalar, min=-1.0, max=1.0)` turns `2` into `1`.
- **`euler_from_quaternion`** — `euler_from_quaternion(*(2*q))` returns `[1.656, 1.571, 2.105]` where `q` gives `[0.3, 0.7, 1.1]`. The `pitch` even saturates at exactly `pi/2` because `sinp = 2*(w*y − z*x)` overshoots 1 and is clamped.

Contrast the two that *are* safe: `quaternion_to_rotation_matrix` calls `normalize_quaternion` internally, and `quaternion_to_axis_angle` is scale-invariant by construction (`atan2` is homogeneous and the `k = 2θ/||v||` factor compensates). Both return bit-identical output for `q` and `2q`.

The sources:

```python
# quaternion_exp_to_log, kornia/geometry/conversions.py:846-854
quaternion_vector = quaternion[..., 1:4]
norm_q = torch.norm(quaternion_vector, p=2, dim=-1, keepdim=True).clamp(min=eps)
quaternion_log = quaternion_vector * torch.acos(torch.clamp(quaternion_scalar, min=-1.0, max=1.0)) / norm_q

# euler_from_quaternion, kornia/geometry/conversions.py:948-956
sinr_cosp = 2.0 * (w * x + y * z)
cosr_cosp = 1.0 - 2.0 * (x * x + yy)
roll = sinr_cosp.atan2(cosr_cosp)
sinp = 2.0 * (w * y - z * x)
sinp = sinp.clamp(min=-1.0, max=1.0)
pitch = sinp.asin()
```

Both formulas are the standard unit-quaternion identities; neither is homogeneous in `q`, and neither function normalizes.

## Repro

```python
import torch
from kornia.geometry.conversions import (
    euler_from_quaternion as e_from_q,
    normalize_quaternion as nq,
    quaternion_exp_to_log as e2l,
    quaternion_from_euler as q_from_e,
    quaternion_to_axis_angle as q2aa,
    quaternion_to_rotation_matrix as q2R,
)

q = torch.tensor([0.5, 0.5, 0.0, 0.0], dtype=torch.float64)   # ||q|| = 0.7071, not unit
print("quaternion_exp_to_log(q)     =", e2l(q).tolist())
print("quaternion_exp_to_log(nq(q)) =", e2l(nq(q)).tolist(), " <- correct")
print("quaternion_exp_to_log([2,0,0,0]) =", e2l(torch.tensor([2.0, 0.0, 0.0, 0.0], dtype=torch.float64)).tolist(), " <- w silently clamped to 1")

qe = torch.stack(q_from_e(*[torch.tensor(a, dtype=torch.float64) for a in (0.3, 0.7, 1.1)]))
print("\neuler_from_quaternion(q)   =", [float(a) for a in e_from_q(*qe)])
print("euler_from_quaternion(2q)  =", [float(a) for a in e_from_q(*(2 * qe))], " <- should be the same rotation")

print("\n# the scale-safe siblings, same 2q input:")
print("quaternion_to_rotation_matrix: max|q2R(2q) - q2R(q)| =", (q2R(2 * qe) - q2R(qe)).abs().max().item())
print("quaternion_to_axis_angle     : max|q2aa(2q) - q2aa(q)| =", (q2aa(2 * qe) - q2aa(qe)).abs().max().item())
```

## Observed

```
quaternion_exp_to_log(q)     = [1.0471975511965976, 0.0, 0.0]
quaternion_exp_to_log(nq(q)) = [0.7853981633974484, 0.0, 0.0]  <- correct
quaternion_exp_to_log([2,0,0,0]) = [0.0, 0.0, 0.0]  <- w silently clamped to 1

euler_from_quaternion(q)   = [0.2999999999999999, 0.6999999999999998, 1.0999999999999999]
euler_from_quaternion(2q)  = [1.6560585860248003, 1.5707963267948966, 2.1048169977173687]  <- should be the same rotation

# the scale-safe siblings, same 2q input:
quaternion_to_rotation_matrix: max|q2R(2q) - q2R(q)| = 0.0
quaternion_to_axis_angle     : max|q2aa(2q) - q2aa(q)| = 0.0
```

## Expected

`q` and `2q` denote the same rotation, so `quaternion_exp_to_log` and `euler_from_quaternion` should return the same answer for both — as `quaternion_to_rotation_matrix` and `quaternion_to_axis_angle` already do in the same file (`0.0` difference on the last two lines). Either:

- **normalize the input** (one call to `normalize_quaternion`, matching `quaternion_to_rotation_matrix`), which also removes the two silent `clamp`s that are currently masking the out-of-range values; or
- **validate** that `||q|| ≈ 1` and raise otherwise, so at least the failure is loud.

The current behavior is the worst of the three: the wrong answer is returned with no signal. It is particularly hazardous because non-unit quaternions arise naturally — from network outputs before a normalization layer, from accumulated products in float32, from interpolation — and because the two `clamp`s make the output *look* plausible (a pitch of exactly `pi/2` reads as a legitimate gimbal-lock result rather than as evidence that the input was rejected in silence).

This is a pure API-consistency defect: three of the five quaternion consumers in this file are scale-safe and two are not, with nothing in any docstring saying which is which.

## Classification

**Semantic / default change.** The behavior on unit quaternions — the documented domain — does not change at all, so a normalizing fix is byte-identical on every in-contract input. But it does change observable output for out-of-contract inputs that currently run without error, and a caller could conceivably be relying on `quaternion_exp_to_log`'s non-homogeneous scaling. Per the reference-core lifecycle (Part I / Part III §2) my reading is that this is close enough to the broken-output class to land early (the affected inputs produce answers that are wrong under any convention), but because it is technically a default change for accepted inputs I am flagging the choice rather than deciding it.

Note that the *gimbal-lock* failure of `euler_from_quaternion` is a separate defect with a separate fix, filed separately; that one occurs with perfectly unit input.

## Environment

kornia 0.9.0rc1 @ `3358ef16`, torch 2.9.1, cpu. Executed from a worktree carrying the batch-3a docs/tests commits on top of `3358ef16`; `quaternion_exp_to_log` and `euler_from_quaternion` are byte-identical to `origin/main` there (no diff hunk lands in either; function bodies compare AST-identical).

Posted on behalf of @ducha-aiki by Claude (Fable 5).
